### PR TITLE
VPN-3381 - Expose vpnglean APIs to QML

### PR DESCRIPTION
**Pull request by brizental**

### CODE REVIEW

File: src/glean/private/event.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+EventMetric::EventMetric(int aId) : m_id(aId) {}
+
+void EventMetric::record() const { glean_event_record_no_extra(m_id); }
+
+void EventMetric::record(QJsonObject extras) {
+  const char* extraValues[extras.size() + 1];
+  const char* extraKeys[extras.size() + 1];
+  int count = 0;
+
+  m_keepQMLStringsAlive.clear();
```
Comment (bakulf) [inline src/glean/private/event.cpp:27]: In this case, you can simply use a `QStringList keepQMLStringsAlive;`

File: src/glean/private/event.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      extraKeys[count] = k.constData();
+      count++;
+    }
+  }
+
+  glean_event_record(m_id, extraKeys, extraValues, count);
+}
+
+void EventMetric::record(EventMetricExtra extras) const {
+  auto ffiExtras = extras.ToFfiExtra();
```
Comment (bakulf) [inline src/glean/private/event.cpp:46]: Another options is to receive a `QStringList&` argument. In this way you can do:

```
QStringList keepAlive;
auto ffiExtras = extras.toFfiExtra(keepAlive);
glean_event_record(m_id, ffiExtras.keys, ffiExtras.values, ffiExtras.count);
```

**Review by bakulf — APPROVED** [review]

File: src/glean/private/event.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#include <QPair>
+#if defined(UNIT_TEST)
+#  include <QJsonDocument>
+#endif
+
+EventMetric::EventMetric(int aId) : m_id(aId) {}
+
+void EventMetric::record() const { glean_event_record_no_extra(m_id); }
+
+void EventMetric::record(QJsonObject extras) {
```
Comment (bakulf) [inline src/glean/private/event.cpp:22]: `const QJsonObject& extras`

File: src/glean/private/event.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  const char* extraValues[extras.size() + 1];
```
Comment (bakulf) [inline src/glean/private/event.cpp:23]: no +1

File: src/glean/private/event.h [hunk 5]
```diff
@@ -6,45 +6,56 @@
...
 
-  void record(T extras) const {
-    auto ffiExtras = extras.ToFfiExtra();
-    glean_event_record(m_id, ffiExtras.keys, ffiExtras.values, ffiExtras.count);
-  }
+  // This function should only be used from QML,
+  // on C++ the variant that receives the FFI extra struct is preferred.
+  //
+  // Note: template classes cannot be used on methods exposed to QML.
+  Q_INVOKABLE void record(QJsonObject extras);
```
Comment (bakulf) [inline src/glean/private/event.h:42]: `const QJsonObject& extras`

File: src/glean/private/event.h [hunk 6]
```diff
@@ -6,45 +6,56 @@
...
 
-#if defined(UNIT_TEST)
-  int32_t testGetNumRecordedErrors(Glean::ErrorType errorType) const {
-    return glean_event_test_get_num_recorded_errors(
-        m_id, static_cast<int32_t>(errorType));
-  }
+  void record(EventMetricExtra extras) const;
```
Comment (bakulf) [inline src/glean/private/event.h:44]: `const EventMetricExtra& extras`

File: src/glean/private/event.h [hunk 7]
```diff
@@ -6,45 +6,74 @@
...
+  //
+  // Note: template classes cannot be used on methods exposed to QML.
+  Q_INVOKABLE void record(const QJsonObject& extras);
 
-#if defined(UNIT_TEST)
-  int32_t testGetNumRecordedErrors(Glean::ErrorType errorType) const {
-    return glean_event_test_get_num_recorded_errors(
-        m_id, static_cast<int32_t>(errorType));
-  }
+  void record(void* extras) const;
```
Comment (brizental) [inline src/glean/private/event.h:61]: Note: this did not work.

File: src/glean/private/event.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      extraValues[count] = v.constData();
+      extraKeys[count] = k.constData();
+      count++;
+    }
+  }
+
+  glean_event_record(m_id, extraKeys, extraValues, count);
+}
+
+void EventMetric::record(std::any extras) const {
```
Comment (brizental) [inline src/glean/private/event.cpp:46]: @bakulf is `std::any` a terrible idea?

File: src/glean/glean.h [hunk 9]
```diff
@@ -2,24 +2,24 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-#ifndef GLEAN_H
-#define GLEAN_H
+#ifndef NEW_GLEAN_H
```
Comment (bakulf) [inline src/glean/glean.h:5]: VPNGLEAN_H?

File: src/glean/private/event.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    : m_id(aId), m_extrasId(extrasId) {}
+
+void EventMetric::record() const { glean_event_record_no_extra(m_id); }
+
+void EventMetric::record(const QJsonObject& extras) {
+  const char* extraValues[extras.size()];
+  const char* extraKeys[extras.size()];
+  int count = 0;
+
+  m_keepStringsAlive.clear();
```
Comment (bakulf) [inline src/glean/private/event.cpp:30]: This can be a QStringList keepStringsAlive

File: src/glean/private/event.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+  foreach (const QString& key, extras.keys()) {
+    auto value = extras.value(key).toString();
```
Comment (bakulf) [inline src/glean/private/event.cpp:33]: What does it happen if this is not a string?

File: src/glean/private/event.cpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      extraKeys[count] = k.constData();
+      count++;
+    }
+  }
+
+  glean_event_record(m_id, extraKeys, extraValues, count);
+}
+
+void EventMetric::record(std::any extras) {
+  EventMetricExtra parsedExtras = std::any_cast<EventMetricExtra>(extras);
```
Comment (bakulf) [inline src/glean/private/event.cpp:49]: In this way you copy data.

File: src/glean/private/event.cpp [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+  Q_ASSERT(parsedExtras.__PRIVATE__id == m_extrasId);
+
+  FfiExtra ffiExtras = parsedExtras.ToFfiExtra(m_keepStringsAlive);
```
Comment (bakulf) [inline src/glean/private/event.cpp:52]: Also here you can use a simple QStringList.

File: src/glean/private/event.cpp [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      extraValues[count] = v.constData();
+      extraKeys[count] = k.constData();
+      count++;
+    }
+  }
+
+  glean_event_record(m_id, extraKeys, extraValues, count);
+}
+
+void EventMetric::record(std::any extras) {
```
Comment (bakulf) [inline src/glean/private/event.cpp:48]: I'm not super happy about this std::any

File: src/glean/private/ping.h [hunk 15]
```diff
@@ -5,13 +5,16 @@
...
 
-#include "vpnglean.h"
+#include <QObject>
 
 class Ping final {
+  Q_GADGET
+
  public:
-  constexpr explicit Ping(int aId) : m_id(aId) {}
+  Ping(int aId);
```
Comment (bakulf) [inline src/glean/private/ping.h:14]: Explicit

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/glean/private/event.cpp (hunk 14 above)
Comment (brizental) [reply src/glean/private/event.cpp:48]: Yes I figured it was probably not good. What do you suggest though?

File: src/glean/private/event.cpp (hunk 11 above)
Comment (brizental) [reply src/glean/private/event.cpp:33]: Hm, that is a good point. I revisited the documentation and contrary to my initial belief this won't turn numbers and booleans into Strings... Nice catch!

File: src/glean/private/event.h [hunk 16]
```diff
@@ -5,46 +5,88 @@
...
+  Q_INVOKABLE QJsonArray testGetValue(const QString& pingName = "") const;
 #endif
 
  private:
   int m_id;
+  EventMetricExtraParser m_parser;
+
+  // Helper vector to extend the lifetime of the strings
+  // that hold the extra key values until they are used.
+  QList<QByteArray> m_keepStringsAlive;
```
Comment (bakulf) [inline src/glean/private/event.h:89]: You can probably remove this.

File: src/glean/private/event.cpp [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    : m_id(aId), m_parser(aParser) {}
+
+void EventMetric::record() const {
+#if not(defined(MVPN_WASM) || defined(BUILD_QMAKE))
+  glean_event_record_no_extra(m_id);
+#endif
+}
+
+void EventMetric::record(const QJsonObject& extras) {
+  FfiExtra ffiExtras = m_parser.fromJsonObject(extras, m_keepStringsAlive);
```
Comment (bakulf) [inline src/glean/private/event.cpp:29]:
```
QStringList keepAStringsAlive;
FfiExtra ffiExtras = m_parser.fromJsonObject(extras, keepsStringsAlive);
```

File: src/glean/private/event.cpp [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+#if not(defined(MVPN_WASM) || defined(BUILD_QMAKE))
+    glean_event_record(m_id, ffiExtras.keys, ffiExtras.values, ffiExtras.count);
+#endif
+  }
+}
+
+void EventMetric::record(EventMetricExtra extras) {
+  Q_ASSERT(extras.__PRIVATE__id == m_id);
+
+  FfiExtra ffiExtras = m_parser.fromStruct(extras, m_keepStringsAlive);
```
Comment (bakulf) [inline src/glean/private/event.cpp:40]: same here

File: src/glean/private/event.h [hunk 19]
```diff
@@ -5,46 +5,88 @@
...
+  }
+};
+
 class EventMetric final {
+  Q_GADGET
+
  public:
-  constexpr explicit EventMetric(int aId) : m_id(aId) {}
+  explicit EventMetric(
+      int aId, EventMetricExtraParser parser = EventMetricExtraParser());
```
Comment (bakulf) [inline src/glean/private/event.h:65]: aId -> id

**Review by bakulf — APPROVED** [review]

**Pull request merged**